lnrpc/routerrpc/router_server.go

```diff
@@ -656,3 +669,192 @@ func (s *Server) SubscribeHtlcEvents(req *SubscribeHtlcEventsRequest,
+	forwardHandler := func(inKey channeldb.CircuitKey,
+		htlc lnwire.UpdateAddHTLC) bool {
+
+		respChan := make(chan bool, 1)
+		newRequest := &htlcInterceptInfo{
+			inKey:        inKey,
+			htlc:         htlc,
+			responseChan: respChan,
+		}
+
+		// timeout is the time after which ForwardHtlcInterceptRequest expire.
+		timeout := time.After(defaultForwardInterceptionTimeout)
+
+		// Send the request to the newRequests channel.
+		select {
+		case newRequests <- newRequest:
+		case <-timeout:
+			log.Errorf("Forward handler returned false - reached timeout of %d",
+				defaultForwardInterceptionTimeout)
+			return false
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+
+		// Receive the response and return it. If no response has been received
+		// in defaultForwardInterceptionTimeout, then return false.
+		select {
+		case resp := <-respChan:
+			return resp
+		case <-timeout:
+			log.Errorf("Forward handler returned false - reached timeout of %d",
+				defaultForwardInterceptionTimeout)
+			return false
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+	}
+
+	// Add our forwardHandler to the Middleware.
+	interceptMiddleware := s.cfg.RouterBackend.HtlcInterceptMiddleware
+	handlerID := interceptMiddleware.AddForwardHtlcHandler(forwardHandler)
+	defer interceptMiddleware.RemoveForwardHtlcHandler(handlerID)
+
+	// errChan is used by the receive loop to signal any errors that occur
+	// during reading from the stream. This is primarily used to shutdown the
+	// send loop in the case of an RPC client disconnecting.
+	errChan := make(chan error, 1)
+
+	// Read user responses and dispatch them to the responses channel.
+	go func() {
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			// Now that we have the response from the RPC client, send it to
+			// the responses chan.
+			select {
+			case responses <- *resp:
+			case <-quit:
+				return
+			case <-s.quit:
+				return
+			}
+		}
+	}()
+
+	// Read both dispatched requests and responses.
+	// For every request send it over the user rpc stream to reach the caller.
+	// For every response coming back from the RPC stream signal its assigned
+	// response channel so the switch will have its result back.
+	pendingRequests := make(map[channeldb.CircuitKey]chan bool)
+	for {
+		select {
+		case newRequest := <-newRequests:
+
+			htlc := newRequest.htlc
+			inKey := newRequest.inKey
+			pendingRequests[inKey] = newRequest.responseChan
+
+			// A ChannelAcceptRequest has been received, send it to the client.
+			interceptionRequest := &ForwardHtlcInterceptRequest{
+				CircuitKey: &CircuitKey{
+					IncomingChanId: inKey.ChanID.ToUint64(),
+					HtlcId:         inKey.HtlcID,
+				},
+				HtlcPaymentHash: htlc.PaymentHash[:],
+				AmountSat:       uint64(htlc.Amount.ToSatoshis()),
+				Expiry:          htlc.Expiry,
+			}
+
+			if err := stream.Send(interceptionRequest); err != nil {
```

Comment:
I understand the concern here.
Interceptions always delegate some responsibility to the user and if we want to mitigate bad use of this API it will come with the price of complicating the switch implementation which I am not sure worth it.
I tend to think that the way to handle this is to use small timeout by default and document it to express this concern.
The API should be used with care and the caller must have all the information ready for the decision on whether to intercept and avoid depending on external resources that may block.
What do you think?

Reply:
I was thinking about using a buffered `Queue`. Could that work? So that the main loop can keep pushing htlcs into the queue and move on.

Reply:
So the concern is not only blocking htlc handling system-wide but also blocking other channels that are polled in the main loop right?
Buffered queue can help here unless the system load is such that the queue ends up full and this also means that these buffered packets need to be re-send to the main loop later.
If that is the case why not introduce a separate channel for fresh forwarded htlcs that will be polled in a separate go routine? 
This go routine will only handle the interception without blocking the main loop.
After the interception handling the packets will be routed to the main loop to be processed regularly as today.

Reply:
Alternatively we can add the interception code in `route` & `routeAsync` which are the only ones that send forwards to the main loop and are executed in the caller go routine.

Reply:
I think it is the same thing, blocking channel receives in the main loop is blocked htlc handling system-wide. In the PR as is, it seems that the main loop can be blocked on a network send to the client. Hand-off to a queue can disconnect the main loop from the rpc code. With queue I don't mean a regular queue, but `queue.ConcurrentQueue`. That already contains its own goroutine, which I think is also what you are suggesting with separate channel.

Reply:
@joostjager I pushed an alternative implementation that introduces a very minimal change to the switch and uses a kind of Proxy structure that wraps the switch and intercepts requests.

In fact the only change to the switch is because settled/failed packets that are forwarded from the interceptor don't have an open circuit which the code expects.
I think conceptually it is not a problem but as part of reviewing this approach we need to find the right solution for that as I am not sure if the current change doesn't comes with the tradeoff of not filtering duplicate settles/failures. 

The blocking issue is now solved without the need for buffering since the ForwardPackets interface method is supposed to be called (and called) in its own go routine.

Reply:
>In fact the only change to the switch is because settled/failed packets that are forwarded from the interceptor don't have an open circuit which the code expects.

Isn't it possible to also not let the return path flow through the switch? So return the failure directly from the `ForwardInterceptor` to the link?

> The blocking issue is now solved without the need for buffering since the ForwardPackets interface method is supposed to be called (and called) in its own go routine.

The only thing remaining is that one packet in the forwarding batch can block the others. But that is much better than a system-wide block that was introduced in the prior version.

Reply:
@joostjager I just pushed another version where the fails/settles from the interceptor are delivered directly to the link using the `mailOrchestrator`
This bypass the attempt to close the circuit that doesn't exist.

A few more notes:
1. For fail packets the switch attempt to encode the error which we don't populate at the moment so we don't loos anything. If in a later time we want to add a failure reason then we have the obfuscator so we can do that in the interceptor.

2. For settled packets the switch log the forward events so if this doesn't count as forward event then all is good, otherwise it is fairly easy to add this logging.

3. It looks like the `mailOrchestrator` is built for concurrent use so it seems OK to use it from the interceptor.

4. Since we isolated the blocking issue to only affect the intercepted packets and not system wide, I tend to think that we don't need to add more concurrency and it is the user responsibility to make sure the packet he intercepts do not block the others. The caller can always handle these packets in an asynchronous way.

Reply:
Looks good, cleaner layering.

1. But we need to do that already in this PR, right? Otherwise what will the sender of the payment see as a failure message?

2. Sounds correct. No forward happened and injecting an external preimage is a special case

3. It may be better to reference the orchestrator as a separate dependency (not via switch), but that is mostly theoretical as there is already a reference to the switch. I am still thinking a little about how this compares to handing the packet off to the link directly (assuming the link is modified to also accept direct packets, like in #4183)

4. Yes it is much safer now. But maybe still need to look at some of the failure modes like network connection lost. Does a timeout still apply?

Reply:
> 1. But we need to do that already in this PR, right? Otherwise what will the sender of the payment see as a failure message?

What do you think is the message type that needed to be sent? Is there anything that supports custom error message?


> 3\. I am still thinking a little about how this compares to handing the packet off to the link directly (

To be honest I think that from the same reason we didn't want the link to talk directly to the interceptor it makes sense for the interceptor to talk to the transport layer rather than find the link itself.

> 4\. But maybe still need to look at some of the failure modes like network connection lost. Does a timeout still apply?

You mean timeout in settle/fail packets? 
As far as I understand the link will re-send packets on startup and the interceptor can/should retry upon error from the orchestrator. Any specific use case you had in mind?

Reply:
>What do you think is the message type that needed to be sent? Is there anything that supports custom error message?

I would think `FailTemporaryChannelFailure`?

>To be honest I think that from the same reason we didn't want the link to talk directly to the interceptor it makes sense for the interceptor to talk to the transport layer rather than find the link itself.

I have no strong opinion about it. Just wondering how the "reliable" mailbox (as long as no restarts) vs direct hand off would compare.

>You mean timeout in settle/fail packets? As far as I understand the link will re-send packets on startup and the interceptor can/should retry upon error from the orchestrator. Any specific use case you had in mind?

I mean what happens if the connection to an application is lost and the packet that is considered for forwarding doesn't get a response over the rpc. Also no connection lost, just nothing happens. And then if the application comes back online, what state does it need to signal forwarding for the 'hanging' forward request?

Reply:
> Just wondering how the "reliable" mailbox (as long as no restarts) vs direct hand off would compare

Using the mailbox let me avoid handling the case where the link is down. The mailbox is reliable in a sense that it will attempt to deliver the packet when the link comes live again.

> I would think `FailTemporaryChannelFailure`?

Added encryption.

Reply:
And if it intercepts, but nothing is ever heard back? I guess that is the application's responsibility then. Similar to how hodl invoices can currently cause channels to close if no resolution is given